fe2-android/app/src/main/java/com/github/dedis/student20_pop/model/data/LAORepository.java

```diff
@@ -61,119 +62,119 @@
-            @NonNull LAODataSource.Local localDataSource,
-            @NonNull AndroidKeysetManager keysetManager,
-            @NonNull Gson gson) {
-        mRemoteDataSource = remoteDataSource;
-        mLocalDataSource = localDataSource;
-        mKeysetManager = keysetManager;
-        mGson = gson;
+  private LAORepository(
+      @NonNull LAODataSource.Remote remoteDataSource,
+      @NonNull LAODataSource.Local localDataSource,
+      @NonNull AndroidKeysetManager keysetManager,
+      @NonNull Gson gson,
+      @NonNull SchedulerProvider schedulerProvider) {
+    mRemoteDataSource = remoteDataSource;
+    mLocalDataSource = localDataSource;
+    mKeysetManager = keysetManager;
+    mGson = gson;
 
-        laoById = new HashMap<>();
-        messageById = new HashMap<>();
-        subscribeRequests = new HashMap<>();
-        catchupRequests = new HashMap<>();
-        createLaoRequests = new HashMap<>();
+    laoById = new HashMap<>();
+    messageById = new HashMap<>();
+    subscribeRequests = new HashMap<>();
+    catchupRequests = new HashMap<>();
+    createLaoRequests = new HashMap<>();
 
-        unprocessed = PublishSubject.create();
+    unprocessed = PublishSubject.create();
 
-        allLaoSubject = BehaviorSubject.create();
+    allLaoSubject = BehaviorSubject.create();
 
-        upstream = mRemoteDataSource.observeMessage().share();
+    upstream = mRemoteDataSource.observeMessage().share();
 
-        subscribedChannels = new HashSet<>();
-        websocketEvents = mRemoteDataSource.observeWebsocket();
+    subscribedChannels = new HashSet<>();
+    websocketEvents = mRemoteDataSource.observeWebsocket();
 
-        // subscribe to incoming messages and the unprocessed message queue
-        startSubscription();
+    this.schedulerProvider = schedulerProvider;
 
-        subscribeToWebsocketEvents();
-    }
+    // subscribe to incoming messages and the unprocessed message queue
+    startSubscription();
+    subscribeToWebsocketEvents();
+  }
 
-    public static LAORepository getInstance(
-            LAODataSource.Remote laoRemoteDataSource,
-            LAODataSource.Local localDataSource,
-            AndroidKeysetManager keysetManager,
-            Gson gson) {
+  public static LAORepository getInstance(
+      LAODataSource.Remote laoRemoteDataSource,
+      LAODataSource.Local localDataSource,
+      AndroidKeysetManager keysetManager,
+      Gson gson, SchedulerProvider schedulerProvider) {
+    if (INSTANCE == null) {
+      synchronized (LAORepository.class) {
         if (INSTANCE == null) {
-            synchronized (LAORepository.class) {
-                if (INSTANCE == null) {
-                    INSTANCE = new LAORepository(laoRemoteDataSource, localDataSource, keysetManager, gson);
-                }
-            }
+          INSTANCE = new LAORepository(laoRemoteDataSource, localDataSource, keysetManager, gson,
+              schedulerProvider);
         }
-        return INSTANCE;
+      }
     }
```

Comment:
SonarSource seems to suggest something like this:

```suggestion
  public static synchronized LAORepository getInstance(
      LAODataSource.Remote laoRemoteDataSource,
      LAODataSource.Local localDataSource,
      AndroidKeysetManager keysetManager,
      Gson gson, SchedulerProvider schedulerProvider) {
    if (INSTANCE == null) {
          INSTANCE = new LAORepository(laoRemoteDataSource, localDataSource, keysetManager, gson,
              schedulerProvider);
    }
    return INSTANCE;
  }
```